File: src/main/java/kr/mjc/rachel/spring/di/quiz/AppConfig.java
package kr.mjc.rachel.spring.di.quiz;

import kr.mjc.rachel.basics.jdbc.DataSourceFactory;
import kr.mjc.rachel.basics.jdbc.JdbcHelper;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

import javax.sql.DataSource;

/**
 * dataSource와 jdbcHelper는 여기서 @Bean으로 구성하고
 * userDao는 @ComponentScan과 @Component로 구성한다.
 */
@Configuration
@ComponentScan(basePackages = "kr.mjc.rachel.spring.di.quiz")
public class AppConfig {
  @Bean
  public DataSource dataSource() {
    return DataSourceFactory.getDataSource();
  }

  @Bean
  public JdbcHelper jdbcHelper(DataSource dataSource) {
    return new JdbcHelper(dataSource);
  }
}

File: src/main/kotlin/kr/mjc/jacob/spring/jdbc/AppConfig.kt
package kr.mjc.jacob.spring.jdbc

import org.mariadb.jdbc.MariaDbDataSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.PropertySource
import org.springframework.core.env.Environment
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import javax.sql.DataSource

@Configuration
@ComponentScan(basePackages = ["kr.mjc.jacob.spring.jdbc.dao"])
@PropertySource("classpath:db.properties")
open class AppConfig {
  @Autowired lateinit var env: Environment

  @Bean
  open fun dataSource() = MariaDbDataSource(env.getProperty("url"))

  @Bean
  open fun jdbcTemplate(dataSource: DataSource) = JdbcTemplate(dataSource)

  @Bean
  open fun nameParameterJdbcTemplate(jdbcTemplate: JdbcTemplate) =
    NamedParameterJdbcTemplate(jdbcTemplate)
}